src/main/java/com/microsoft/azure/vmagent/AzureVMCloud.java

```diff
@@ -803,79 +813,97 @@ public AzureVMDeploymentInfo call() throws AzureCloudException {
+                                            computer.connect(false).get();
+                                        } else if (agent.getAgentLaunchMethod()
+                                                .equalsIgnoreCase("JNLP")) {
+                                            // Wait until node is online
+                                            waitUntilJNLPNodeIsOnline(agent);
+                                        }
+                                    } finally {
+                                        // Place node in default state, now can be
+                                        // dealt with by the cleanup task.
+                                        agent.clearCleanUpAction();
+                                    }
+                                } catch (Exception e) {
                                     LOGGER.log(
                                             Level.SEVERE,
-                                            String.format("Failure removing Jenkins node for '%s'", vmName),
-                                            nodeRemoveEx);
-                                    // Do not throw to avoid it being recorded
+                                            String.format("Failure to in post-provisioning for '%s'", vmName),
+                                            e);
+
+                                    handleFailure(template, vmName, e, FailureStage.POSTPROVISIONING);
+
+                                    // Remove the node from jenkins
+                                    try {
+                                        Jenkins.getInstance().removeNode(agent);
+                                    } catch (IOException nodeRemoveEx) {
+                                        LOGGER.log(
+                                                Level.SEVERE,
+                                                String.format("Failure removing Jenkins node for '%s'", vmName),
+                                                nodeRemoveEx);
+                                        // Do not throw to avoid it being recorded
+                                    }
+
+                                    throw AzureCloudException.create(e);
+                                }
+
+                                if (isProvisionOutside) {
+                                    CloudStatistics.ProvisioningListener.get().onComplete(provisioningId, agent);
                                 }
 
-                                throw AzureCloudException.create(e);
+                                return agent;
+                            } catch (AzureCloudException e) {
+                                if (isProvisionOutside) {
+                                    CloudStatistics.ProvisioningListener.get().onFailure(provisioningId, e);
+                                }
+                                throw e;
+                            } finally {
+                                PoolLock.provisionUnlock(template);
```

Comment:
This will have issue if multiple thread is provision the same template, the first completed thread will unlock the template.